pkgs/by-name/du/duckstation/package.nix

```diff
@@ -1,35 +1,42 @@
-{ lib
-, stdenv
-, fetchFromGitHub
-, SDL2
-, cmake
-, cubeb
-, curl
-, extra-cmake-modules
-, libXrandr
-, libbacktrace
-, makeWrapper
-, ninja
-, pkg-config
-, qtbase
-, qtsvg
-, qttools
-, qtwayland
-, substituteAll
-, vulkan-loader
-, wayland
-, wrapQtAppsHook
+{
+  lib,
+  SDL2,
+  cmake,
+  cubeb,
+  curl,
+  extra-cmake-modules,
+  fetchFromGitHub,
+  libXrandr,
+  libbacktrace,
+  libwebp,
+  makeWrapper,
+  ninja,
+  pkg-config,
+  qt6,
+  shaderc,
+  stdenv,
+  substituteAll,
+  vulkan-loader,
+  wayland,
 }:
-
+let
+  inherit (qt6)
```

Comment:
I think we should use qt6.callPackage here instead

Reply:
This is not compatible with by-name hierarchy.

Reply:
Avoiding the wrapper by copying what it does into the derivation is definitely not the solution to this.

by-name can have calls from all-packages.nix for exactly those use cases and it is allowed by design.

Reply:
by-name does not accept `qt6.callPackage`, only `callPackage` itself.

Further, to be honest I have problems with those magic `hugePackageSet.callPackage` things, they look too mysterious.
Further, what does this wrapper do that can't be done without it?

Further (I like that word!), I remember very well that some packages had `libsForQt5.callPackage` when it was otherwise not needed. `cmake` was a clear example.

Reply:
Merging is blocked due to this @SuperSandro2000 's review, so I marked this conversation as unresolved. I think it is OK to not use `qt6.callPackage` here, especially if it allows us to transition to `pkgs/by-name`, and I think it should be the choice of the maintainer to use it or not, but that's only my opinion.

Reply:
qt6.callPackage does more than bringing the qt packages into scope. It also does splicing which is complicated (#265524, #269654) and if it is slightly off, it breaks cross compilation. It also uses the correct stdenv for darwin but that is not relevant for this package. Instead of suggesting this pattern to other contributors and luring them into broken darwin and cross builds, please just use the qt6.callPackage.

If other ecosystems have a patterns on how to consume them, we should really use them, especially if it is easily possible, to avoid issues for our packages and to avoid extra work for the ecosystem maintainers and especially avoid extra work debugging hard to debug topics like cross compilation which is especially hairy inside of qt.

pkgs by-name is not a good reason to avoid the wrapper. A good known reason is when qt is used inside another wrapper like buildPythonPackage.

libsForQt5 also generated the package for every minor qt5 version which is right now not relevant and cmake probably doesn't need it but duckstation is I think a qt application.

Reply:
Oh yeah the cross compilation is a good argument. Recently the qt docs [have been updated](https://github.com/NixOS/nixpkgs/pull/287169) by @K900 (also ping @Aleksanaa and @eclairevoyant) to avoid recommending `qt6.callPackage`, but the cross compilation argument never came up.

There is in fact still a way to use `qt6.callPackage` with `pkgs/by-name`, but you need to nest it:

```nix
# pkgs/by-name/fo/foo/package.nix
{ qt6 }:
qt6.callPackage (
  { qtbase, stdenv }:
  stdenv.mkDerivation { ... }
) { }
```

Reply:
> it breaks cross compilation.

Cross-compilation is already broken in this package.

<details>
<summary><code>nix-build -A pkgsCross.aarch64-multiplatform.duckstation --option keep-going false</code></summary>

```
error:
       … while calling the 'derivationStrict' builtin
         at <nix/derivation-internal.nix>:9:12:
            8|
            9|   strict = derivationStrict drvAttrs;
             |            ^
           10|

       … while evaluating derivation 'duckstation-aarch64-unknown-linux-gnu-0.1-6658'
         whose name attribute is located at /mnt/scratch/nixpkgs/pkgs/stdenv/generic/make-derivation.nix:331:7

       … while evaluating attribute 'postFixup' of derivation 'duckstation-aarch64-unknown-linux-gnu-0.1-6658'
         at /mnt/scratch/nixpkgs/pkgs/by-name/du/duckstation/package.nix:119:3:
          118|   # Can't avoid the double wrapping, the binary wrapper from qtWrapperArgs doesn't support --run
          119|   postFixup = ''
             |   ^
          120|     source "${makeWrapper}/nix-support/setup-hook"

       (stack trace truncated; use '--show-trace' to show the full, detailed trace)

       error: makeWrapper/makeShellWrapper must be in nativeBuildInputs
```

</details>

After fixing with the following diff:

<details>

```diff
diff --git a/pkgs/by-name/du/duckstation/package.nix b/pkgs/by-name/du/duckstation/package.nix
index 37d2d01534b1..3c29c8f0daec 100644
--- a/pkgs/by-name/du/duckstation/package.nix
+++ b/pkgs/by-name/du/duckstation/package.nix
@@ -16,18 +16,16 @@
 , qt6
 , substituteAll
 , vulkan-loader
+, qtbase
+, qtsvg
+, qttools
+, qtwayland
+, wrapQtAppsHook
 , wayland
 }:
 
 let
   shaderc-patched = callPackage ./shaderc-patched.nix { };
-  inherit (qt6)
-    qtbase
-    qtsvg
-    qttools
-    qtwayland
-    wrapQtAppsHook
-  ;
 in
 stdenv.mkDerivation (finalAttrs: {
   pname = "duckstation";
@@ -60,6 +58,7 @@ stdenv.mkDerivation (finalAttrs: {
     pkg-config
     qttools
     wrapQtAppsHook
+    makeWrapper
   ];
 
   buildInputs = [
@@ -118,7 +117,6 @@ stdenv.mkDerivation (finalAttrs: {
   # https://github.com/stenzek/duckstation/blob/master/scripts/appimage/apprun-hooks/default-to-x11.sh
   # Can't avoid the double wrapping, the binary wrapper from qtWrapperArgs doesn't support --run
   postFixup = ''
-    source "${makeWrapper}/nix-support/setup-hook"
     wrapProgram $out/bin/duckstation-qt \
       --run 'if [[ -z $I_WANT_A_BROKEN_WAYLAND_UI ]]; then export QT_QPA_PLATFORM=xcb; fi'
   '';
diff --git a/pkgs/top-level/all-packages.nix b/pkgs/top-level/all-packages.nix
index a4d7e00bd616..a485289fb4bb 100644
--- a/pkgs/top-level/all-packages.nix
+++ b/pkgs/top-level/all-packages.nix
@@ -41039,4 +41039,6 @@ with pkgs;
   dillo = callPackage ../by-name/di/dillo/package.nix {
     fltk = fltk13;
   };
+
+  duckstation = qt6.callPackage ../by-name/du/duckstation/package.nix { };
 }
```
</details>

 we get

<details>

```
these 26 derivations will be built:
  /nix/store/c0sndkqwvr8v15qbg7nl6in4pkxdp8xz-psqlodbc-aarch64-unknown-linux-gnu-10.01.0000.drv
  /nix/store/p7axjbgyy75k6971ccg9lyyl7swggx0y-qtbase-aarch64-unknown-linux-gnu-6.7.1.drv
  /nix/store/28mzzsxwsnzagg1phy7wzk747fmqz0b5-qtlanguageserver-aarch64-unknown-linux-gnu-6.7.1.drv
  /nix/store/49bf87db3pxd61523jhv9wzv1pwxx0ih-qtsvg-aarch64-unknown-linux-gnu-6.7.1.drv
  /nix/store/g89cpcfa3dq4pwk0wp6vq5fh9lx22pcx-qttranslations-6.7.1.drv
  /nix/store/8mwdxyj4k01m9bskcqw5bd0iydn280vp-qtbase-6.7.1.drv
  /nix/store/4nr0ia7yg5f5iq6bjwnvj3jxpwn7qfk1-qtlanguageserver-6.7.1.drv
  /nix/store/8lpsajmwdvwq0cdmwaj2aw2l2rznz4z4-qtsvg-aarch64-unknown-linux-gnu-5.15.12.drv
  /nix/store/b9rvq2zw7qzfifqqwc82rsf7s7jvvq2y-qtdeclarative-aarch64-unknown-linux-gnu-5.15.12.drv
  /nix/store/5v5ha9yn3jqi9r9x2w0psxbnqdljjldb-qtwebchannel-aarch64-unknown-linux-gnu-5.15.12.drv
  /nix/store/krgycar52dh123x4c483bfw1rvsj7bm5-python3.11-pyqt5-5.15.9-aarch64-unknown-linux-gnu.drv
  /nix/store/68ryybbx4z7fyd1kzszjz2skciw19p9x-python3-aarch64-unknown-linux-gnu-3.11.9-env.drv
  /nix/store/mspls9azizz05d4rcvhisg1yz7brq9jk-qtsvg-5.15.12.drv
  /nix/store/7ar0rgvx6hx6ng5z7iw2d45vi61m1qj1-qtdeclarative-5.15.12.drv
  /nix/store/xn1970m3bw410lxiicbf9125grhmhkzj-qtshadertools-aarch64-unknown-linux-gnu-6.7.1.drv
  /nix/store/n2xiqaya1y12hx6riif63pgdn2bvv55p-qtdeclarative-aarch64-unknown-linux-gnu-6.7.1.drv
  /nix/store/b68lsdbi5wny6zgbd31baj08xbx6qrfx-qtwayland-aarch64-unknown-linux-gnu-6.7.1.drv
  /nix/store/d5dgjms7mkc44y8qyynbrcaqcgij9v8i-qtquickcontrols-5.15.12.drv
  /nix/store/8zsw31dyizv9bp5xg3a0qq47kf1sxc2f-qtwayland-5.15.12.drv
  /nix/store/mk4q210p59fwp99ralhdhg5kjd2mn7pn-wrap-qt5-apps-hook.drv
  /nix/store/dbzcs11w17z9pr5hqn0ikk29qrsfv63r-ffado-aarch64-unknown-linux-gnu-2.4.8.drv
  /nix/store/isgcnw2rx2i6ygk2g99lywn0r7k1gq69-jack2-aarch64-unknown-linux-gnu-1.9.22.drv
  /nix/store/ncb6vp6ybwx7igp7sp8ibcmsnvrmxr1a-qtshadertools-6.7.1.drv
  /nix/store/i9r3nchvhsbcjvjm5z9rx7cfczm2q9fh-qtdeclarative-6.7.1.drv
  /nix/store/lihl5w0zl5nrw3kk25pxgl7kafl6nd6l-qttools-6.7.1.drv
  /nix/store/8qd2jxh1v6wf2vr3bhr91m49drpb32i0-duckstation-aarch64-unknown-linux-gnu-0.1-6658.drv
building '/nix/store/c0sndkqwvr8v15qbg7nl6in4pkxdp8xz-psqlodbc-aarch64-unknown-linux-gnu-10.01.0000.drv'...
building '/nix/store/g89cpcfa3dq4pwk0wp6vq5fh9lx22pcx-qttranslations-6.7.1.drv'...
building '/nix/store/8lpsajmwdvwq0cdmwaj2aw2l2rznz4z4-qtsvg-aarch64-unknown-linux-gnu-5.15.12.drv'...
building '/nix/store/mspls9azizz05d4rcvhisg1yz7brq9jk-qtsvg-5.15.12.drv'...
Running phase: unpackPhase
unpacking source archive /nix/store/gq9jk11lkim0fvy504593nd4rkwzv0kr-psqlodbc-10.01.0000.tar.gz
source root is psqlodbc-10.01.0000
setting SOURCE_DATE_EPOCH to timestamp 1514380278 of file psqlodbc-10.01.0000/test/Makefile
Running phase: patchPhase
Running phase: updateAutotoolsGnuConfigScriptsPhase
Updating Autotools / GNU config script to a newer upstream version: ./config/config.sub
Updating Autotools / GNU config script to a newer upstream version: ./config/config.guess
Running phase: updateAutotoolsGnuConfigScriptsPhase
Updating Autotools / GNU config script to a newer upstream version: ./config/config.sub
Updating Autotools / GNU config script to a newer upstream version: ./config/config.guess
Running phase: configurePhase
fixing libtool script ./config/ltmain.sh
patching script interpreter paths in ./configure
./configure: interpreter directive changed from "#! /bin/sh" to "/nix/store/306znyj77fv49kwnkpxmb0j2znqpa8bj-bash-5.2p26/bin/sh"
configure flags: --disable-static --disable-dependency-tracking --prefix=/nix/store/r8bx0c4whyz4ki1bx3ng3l7cnahg7ig6-psqlodbc-aarch64-unknown-linux-gnu-10.01.0000 --build=x86_64-unknown-linux-gnu --host=aarch64-unknown-linux-gnu
checking for a BSD-compatible install... /nix/store/php4qidg2bxzmm79vpri025bqi0fa889-coreutils-9.5/bin/install -c
checking whether build environment is sane... yes
checking for aarch64-unknown-linux-gnu-strip... aarch64-unknown-linux-gnu-strip
checking for a thread-safe mkdir -p... /nix/store/php4qidg2bxzmm79vpri025bqi0fa889-coreutils-9.5/bin/mkdir -p
checking for gawk... gawk
checking whether make sets $(MAKE)... yes
checking whether make supports nested variables... yes
checking whether to enable maintainer-specific portions of Makefiles... no
checking for aarch64-unknown-linux-gnu-gcc... aarch64-unknown-linux-gnu-gcc
Error: detected mismatched Qt dependencies:
    /nix/store/lrz3zjch3xyihm08a3q2p7fin13a169l-qtbase-aarch64-unknown-linux-gnu-5.15.12-dev
    /nix/store/6fr1msliinblzypw9zz51ffib1nzrkd4-qtbase-5.15.12-dev
Error: detected mismatched Qt dependencies:
    /nix/store/6fr1msliinblzypw9zz51ffib1nzrkd4-qtbase-5.15.12-dev
    /nix/store/p8cr0k5g2jbjy6k80sdhfyr7vp4zzl8w-qtbase-5.15.12-dev
error: builder for '/nix/store/8lpsajmwdvwq0cdmwaj2aw2l2rznz4z4-qtsvg-aarch64-unknown-linux-gnu-5.15.12.drv' failed with exit code 1;
       last 3 log lines:
       > Error: detected mismatched Qt dependencies:
       >     /nix/store/lrz3zjch3xyihm08a3q2p7fin13a169l-qtbase-aarch64-unknown-linux-gnu-5.15.12-dev
       >     /nix/store/6fr1msliinblzypw9zz51ffib1nzrkd4-qtbase-5.15.12-dev
       For full logs, run 'nix log /nix/store/8lpsajmwdvwq0cdmwaj2aw2l2rznz4z4-qtsvg-aarch64-unknown-linux-gnu-5.15.12.drv'.
error: 1 dependencies of derivation '/nix/store/krgycar52dh123x4c483bfw1rvsj7bm5-python3.11-pyqt5-5.15.9-aarch64-unknown-linux-gnu.drv' failed to build
error: 1 dependencies of derivation '/nix/store/dbzcs11w17z9pr5hqn0ikk29qrsfv63r-ffado-aarch64-unknown-linux-gnu-2.4.8.drv' failed to build
error: 1 dependencies of derivation '/nix/store/isgcnw2rx2i6ygk2g99lywn0r7k1gq69-jack2-aarch64-unknown-linux-gnu-1.9.22.drv' failed to build
error: 1 dependencies of derivation '/nix/store/8qd2jxh1v6wf2vr3bhr91m49drpb32i0-duckstation-aarch64-unknown-linux-gnu-0.1-6658.drv' failed to build
```

</details>

This error happens whether you switch back to `qt6.callPackage` and switch the expression args or not... so no one is bothering to cross-compile `duckstation` anyway, this argument is pointless. Fixing cross-compilation should be done in another PR.

Reply:
Qt cross is woefully broken, and if it is ever fixed, it should be handled by makeScopeWithSplicing and friends, not whatever this is.

Reply:
> There is in fact still a way to use `qt6.callPackage` with `pkgs/by-name`, but you need to nest it:
> 
> ```nix
> # pkgs/by-name/fo/foo/package.nix
> { qt6 }:
> qt6.callPackage (
>   { qtbase, stdenv }:
>   stdenv.mkDerivation { ... }
> ) { }
> ```

@AndersonTorres this looks like the best option to me - both satisfying @SuperSandro2000's request and the desire to use `pkgs/by-name`.

Reply:
That's as bad as doing `{ pkgs }: pkgs.stdenv.mkDerivation { ... }`, except worse, since you can't even override anything under `pkgs` here.

Rather than tortured solutions to theoretical concerns that haven't even been theoretically demonstrated, let's concretely decide on either `by-name` or `qt6.callPackage`, not both.

Reply:
By the way, using `by-name` does not break splicing; _inheriting_ from `qt6` when `callPackage`-ing _would_ break splicing (see #204303).

To adapt the example shown in that issue:
```nix
# test.nix
let
  pkgs = import <nixpkgs> { };
  pkgsCross = pkgs.pkgsCross.aarch64-multiplatform;
  test1 = pkgsCross.callPackage
    ({ qt6, xorg }: {
      q = qt6.qtbase.__spliced;
      x = xorg.libX11.__spliced;
    })
    { };

  test2 = pkgsCross.callPackage
    ({ qtbase, libX11 }: {
      d = qtbase.__spliced;
      x = libX11.__spliced;
    })
    {
      inherit (pkgsCross.qt6) qtbase;
      inherit (pkgsCross.xorg) libX11;
    };
in
{
  # for --json to work
  workingSplices =
    removeAttrs test1 [ "override" "overrideDerivation" ];
  notWorkingSplices =
    removeAttrs test2 [ "override" "overrideDerivation" ];
}
```


<details>
<summary><code>NIXPKGS_ALLOW_UNSUPPORTED_SYSTEM=1 nix eval -f test.nix workingSplices --json | jq</code></summary>

```json
{
  "q": {
    "buildBuild": "/nix/store/ikqvl8jq0d7vlp25d9zlbh44m4hljahc-qtbase-6.7.1",
    "buildHost": "/nix/store/lfqjxhf4vf6nwz75sjzg1ixqaszzz1ay-qtbase-6.7.1",
    "buildTarget": "/nix/store/lfqjxhf4vf6nwz75sjzg1ixqaszzz1ay-qtbase-6.7.1",
    "hostHost": "/nix/store/b1547ipq4rbzxwccs2x8fh52bw07h8rw-qtbase-aarch64-unknown-linux-gnu-6.7.1",
    "hostTarget": "/nix/store/b1547ipq4rbzxwccs2x8fh52bw07h8rw-qtbase-aarch64-unknown-linux-gnu-6.7.1"
  },
  "x": {
    "buildBuild": "/nix/store/x9fw7rbdb34gq0f8q750kw344lbv9nk1-libX11-1.8.9",
    "buildHost": "/nix/store/x9fw7rbdb34gq0f8q750kw344lbv9nk1-libX11-1.8.9",
    "buildTarget": "/nix/store/x9fw7rbdb34gq0f8q750kw344lbv9nk1-libX11-1.8.9",
    "hostHost": "/nix/store/0ds953302knjk042h2j78pyz851zip5r-libX11-aarch64-unknown-linux-gnu-1.8.9",
    "hostTarget": "/nix/store/0ds953302knjk042h2j78pyz851zip5r-libX11-aarch64-unknown-linux-gnu-1.8.9"
  }
}
```

</details>

<details>
<summary><code>NIXPKGS_ALLOW_UNSUPPORTED_SYSTEM=1 nix eval -f test.nix notWorkingSplices --json | jq</code></summary>

```
error:
       … while evaluating attribute 'd'
         at /home/jelly3781/Desktop/test.nix:13:7:
           12|     ({ qtbase, libX11 }: {
           13|       d = qtbase.__spliced;
             |       ^
           14|       x = libX11.__spliced;

       error: attribute '__spliced' missing
       at /home/jelly3781/Desktop/test.nix:13:11:
           12|     ({ qtbase, libX11 }: {
           13|       d = qtbase.__spliced;
             |           ^
           14|       x = libX11.__spliced;
```

</details>

So I consider this is a non-issue, splicing works.

Reply:
Closing as solved.